### ISSUE TITLE
[Bug]: Flush 120s timeout when enable level zero segment

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#30552]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: master-20240204-18b979d9-amd64
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka): pulsar   
- SDK version(e.g. pymilvus v2.0.0rc2): pymilvus 2.4.0rc24
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1. deploy milvus and wnable L0 segment
2. scene-1 of flush 120s timeout
   - concurrent search + insert + delete + flush [argo: level-zero-stable-13a-1952273556](https://argo-workflows.zilliz.cc/archived-workflows/qa/49d1b28c-5a67-4c14-ba28-f58b7a6d673b?nodeId=level-zero-stable-13a-1952273556)
   - pods:
```
level-zero-insert-13-etcd-0                                       1/1     Running       0               24h
level-zero-insert-13-etcd-1                                       1/1     Running       0               24h
level-zero-insert-13-etcd-2                                       1/1     Running       0               24h
level-zero-insert-13-milvus-datanode-5f45bf7fdb-kvfjn             1/1     Running       0               24h
level-zero-insert-13-milvus-datanode-5f45bf7fdb-vq5f5             1/1     Running       0               24h
level-zero-insert-13-milvus-indexnode-7fb5bc4657-6xcp8            1/1     Running       0               24h
level-zero-insert-13-milvus-indexnode-7fb5bc4657-f59mf            1/1     Running       0               24h
level-zero-insert-13-milvus-mixcoord-694688f8b9-gc9tw             1/1     Running       0               24h
level-zero-insert-13-milvus-proxy-c9f54585-mwq4z                  1/1     Running       0               24h
level-zero-insert-13-milvus-querynode-0-5948f94c49-nds64          1/1     Running       0               24h
level-zero-insert-13-milvus-querynode-0-5948f94c49-rkhtj          1/1     Running       0               24h
level-zero-insert-13-milvus-querynode-0-5948f94c49-vv8zn          1/1     Running       0               24h
level-zero-insert-13-minio-0                                      1/1     Running       0               24h
level-zero-insert-13-minio-1                                      1/1     Running       0               24h
level-zero-insert-13-minio-2                                      1/1     Running       0               24h
level-zero-insert-13-minio-3                                      1/1     Running       0               24h
level-zero-insert-13-pulsar-bookie-0                              1/1     Running       0               24h
level-zero-insert-13-pulsar-bookie-1                              1/1     Running       0               24h
level-zero-insert-13-pulsar-bookie-2                              1/1     Running       0               24h
level-zero-insert-13-pulsar-bookie-init-lmt9q                     0/1     Completed     0               24h
level-zero-insert-13-pulsar-broker-0                              1/1     Running       0               24h
level-zero-insert-13-pulsar-proxy-0                               1/1     Running       0               24h
level-zero-insert-13-pulsar-pulsar-init-725rs                     0/1     Completed     0               24h
level-zero-insert-13-pulsar-recovery-0                            1/1     Running       0               24h
level-zero-insert-13-pulsar-zookeeper-0                           1/1     Running       0               24h
level-zero-insert-13-pulsar-zookeeper-1                           1/1     Running       0               24h
level-zero-insert-13-pulsar-zookeeper-2                           1/1     Running       0               24h
```

3. scene-2 of flush 120s timeout:
   - concurrent search + insert_into_10_partitions + delete + flush [argo: level-zero-stable-13a-347164782](https://argo-workflows.zilliz.cc/archived-workflows/qa/49d1b28c-5a67-4c14-ba28-f58b7a6d673b?nodeId=level-zero-stable-13a-347164782)
   - pods:
```
level-zero-multi-par-13-etcd-0                                    1/1     Running            0                 24h
level-zero-multi-par-13-etcd-1                                    1/1     Running            0                 24h
level-zero-multi-par-13-etcd-2                                    1/1     Running            0                 24h
level-zero-multi-par-13-milvus-datanode-555f7d9876-kjpt8          1/1     Running            0                 24h
level-zero-multi-par-13-milvus-datanode-555f7d9876-pl6fb          1/1     Running            0                 24h
level-zero-multi-par-13-milvus-indexnode-66764db699-m6s7f         1/1     Running            0                 24h
level-zero-multi-par-13-milvus-indexnode-66764db699-wspxt         1/1     Running            0                 24h
level-zero-multi-par-13-milvus-mixcoord-8577f67559-cj89l          1/1     Running            0                 24h
level-zero-multi-par-13-milvus-proxy-d76994f5-d2q4r               1/1     Running            0                 24h
level-zero-multi-par-13-milvus-querynode-0-7b8d45c5b5-b6wb4       1/1     Running            0                 24h
level-zero-multi-par-13-milvus-querynode-0-7b8d45c5b5-jxvcr       1/1     Running            0                 24h
level-zero-multi-par-13-milvus-querynode-0-7b8d45c5b5-r7znh       1/1     Running            0                 24h
level-zero-multi-par-13-minio-0                                   1/1     Running            0                 24h
level-zero-multi-par-13-minio-1                                   1/1     Running            0                 24h
level-zero-multi-par-13-minio-2                                   1/1     Running            0                 24h
level-zero-multi-par-13-minio-3                                   1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-bookie-0                           1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-bookie-1                           1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-bookie-2                           1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-bookie-init-fqfpt                  0/1     Completed          0                 24h
level-zero-multi-par-13-pulsar-broker-0                           1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-proxy-0                            1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-pulsar-init-lzms7                  0/1     Completed          0                 24h
level-zero-multi-par-13-pulsar-recovery-0                         1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-zookeeper-0                        1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-zookeeper-1                        1/1     Running            0                 24h
level-zero-multi-par-13-pulsar-zookeeper-2                        1/1     Running            0                 24h
```

### Expected Behavior

L0 segment compaction does not affect Flush

**Comment by ThreadDao** [comment]
image: master-20240221-77eb6def
Very few flush requests return 120s timeout. Maybe don’t block release first ?
![image](https://github.com/milvus-io/milvus/assets/27288593/daa394b9-7c9d-487f-b25a-aefd6b86dc94)
```
[2024-02-21 09:20:15,177 - WARNING - fouram]: [93mRetry timeout: 60s[0m (decorators.py:100)
[2024-02-21 09:20:15,178 - ERROR - fouram]: RPC error: [flush], <MilvusException: (code=1, message=Retry timeout: 60s, message=wait for flush timeout, collection: fouram_mqiWnSeO)>, <Time:{'RPC start': '2024-02-21 09:19:03.374017', 'RPC error': '2024-02-21 09:20:15.178038'}> (decorators.py:134)
```
- argo: [level-zero-stable-15-1825094311](https://argo-workflows.zilliz.cc/archived-workflows/qa/7d1bf97d-0f5d-4e2e-b70d-ec3b890bdea0?nodeId=level-zero-stable-15-1825094311)
- pods:
```
level-zero-insert-15-etcd-0                                       1/1     Running       0                 8h
level-zero-insert-15-etcd-1                                       1/1     Running       0                 8h
level-zero-insert-15-etcd-2                                       1/1     Running       0                 8h
level-zero-insert-15-milvus-datanode-58d8c66489-cz9q5             1/1     Running       0                 8h
level-zero-insert-15-milvus-datanode-58d8c66489-k4qjk             1/1     Running       0                 8h
level-zero-insert-15-milvus-indexnode-74bf9f4fbd-ln7vm            1/1     Running       0                 8h
level-zero-insert-15-milvus-indexnode-74bf9f4fbd-vx68s            1/1     Running       0                 8h
level-zero-insert-15-milvus-mixcoord-b699b6586-q8hfm              1/1     Running       0                 8h
level-zero-insert-15-milvus-proxy-c9df6dff8-klj8z                 1/1     Running       0                 8h
level-zero-insert-15-milvus-querynode-0-6cd768d9bc-b6jnh          1/1     Running       0                 8h
level-zero-insert-15-milvus-querynode-0-6cd768d9bc-h5kj2          1/1     Running       0                 8h
level-zero-insert-15-milvus-querynode-0-6cd768d9bc-msvq8          1/1     Running       0                 8h
level-zero-insert-15-minio-0                                      1/1     Running       0                 8h
level-zero-insert-15-minio-1                                      1/1     Running       0                 8h
level-zero-insert-15-minio-2                                      1/1     Running       0                 8h
level-zero-insert-15-minio-3                                      1/1     Running       0                 8h
level-zero-insert-15-pulsar-bookie-0                              1/1     Running       0                 8h
level-zero-insert-15-pulsar-bookie-1                              1/1     Running       0                 8h
level-zero-insert-15-pulsar-bookie-2                              1/1     Running       0                 8h
level-zero-insert-15-pulsar-bookie-init-6btlx                     0/1     Completed     0                 8h
level-zero-insert-15-pulsar-broker-0                              1/1     Running       0                 8h
level-zero-insert-15-pulsar-proxy-0                               1/1     Running       0                 8h
level-zero-insert-15-pulsar-pulsar-init-vnwc6                     0/1     Completed     0                 8h
level-zero-insert-15-pulsar-recovery-0                            1/1     Running       0                 8h
level-zero-insert-15-pulsar-zookeeper-0                           1/1     Running       0                 8h
level-zero-insert-15-pulsar-zookeeper-1                           1/1     Running       0                 8h
level-zero-insert-15-pulsar-zookeeper-2                           1/1     Running       0                 8h
```

**Comment by ThreadDao** [comment]
Multiple partitions have the same problem. Very few partition.flush requests return 120s timeout, delete and search deadline exceed. Maybe don’t block release first ?
![image](https://github.com/milvus-io/milvus/assets/27288593/86beba9c-584c-4071-b820-af9ed846d17e)

image: master-20240221-77eb6def

- argo: [level-zero-stable-15-par](https://argo-workflows.zilliz.cc/archived-workflows/qa/1cd31549-6d03-4bea-a0c5-719937812cec?nodeId=level-zero-stable-15-par)
- pods:
```
level-zero-multi-par-15-etcd-0                                    1/1     Running       0                 8h
level-zero-multi-par-15-etcd-1                                    1/1     Running       0                 8h
level-zero-multi-par-15-etcd-2                                    1/1     Running       0                 8h
level-zero-multi-par-15-milvus-datanode-6d44949959-jclls          1/1     Running       0                 8h
level-zero-multi-par-15-milvus-datanode-6d44949959-mngcc          1/1     Running       0                 8h
level-zero-multi-par-15-milvus-indexnode-869f696774-rb5bc         1/1     Running       0                 8h
level-zero-multi-par-15-milvus-indexnode-869f696774-rpqm9         1/1     Running       0                 8h
level-zero-multi-par-15-milvus-mixcoord-6486664c6d-hwgp7          1/1     Running       0                 8h
level-zero-multi-par-15-milvus-proxy-75bcf54bd4-5wvf6             1/1     Running       0                 8h
level-zero-multi-par-15-milvus-querynode-0-68f47c899f-82b7d       1/1     Running       0                 8h
level-zero-multi-par-15-milvus-querynode-0-68f47c899f-9qgjg       1/1     Running       0                 8h
level-zero-multi-par-15-milvus-querynode-0-68f47c899f-bvptb       1/1     Running       0                 8h
level-zero-multi-par-15-minio-0                                   1/1     Running       0                 8h
level-zero-multi-par-15-minio-1                                   1/1     Running       0                 8h
level-zero-multi-par-15-minio-2                                   1/1     Running       0                 8h
level-zero-multi-par-15-minio-3                                   1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-bookie-0                           1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-bookie-1                           1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-bookie-2                           1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-bookie-init-9vt25                  0/1     Completed     0                 8h
level-zero-multi-par-15-pulsar-broker-0                           1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-proxy-0                            1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-pulsar-init-knnq6                  0/1     Completed     0                 8h
level-zero-multi-par-15-pulsar-recovery-0                         1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-zookeeper-0                        1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-zookeeper-1                        1/1     Running       0                 8h
level-zero-multi-par-15-pulsar-zookeeper-2                        1/1     Running       0                 8h
```

**Comment by xiaofan-luan** [comment]
we still need to figure out the reason since 120s is a very long time.

Might be related to a issue @xiaocai2333 is working on , flush is block by compaction,

**Comment by XuanYang-cn** [comment]
when l0 enabled, compaction shouldn't block flush

**Comment by ThreadDao** [comment]
update:
- image: master-20240307-ba8197fa-amd64
- argo: https://argo-workflows.zilliz.cc/archived-workflows/qa/9e624fb9-d85e-4a57-b7ce-43c2d6de088c?nodeId=level-zero-mem-1a-1664299480
- 3 flush 60s timeout [Loki client logs](https://grafana-4am.zilliz.cc/explore?panes=%7B%22Qa9%22:%7B%22datasource%22:%22vhI6Vw67k%22,%22queries%22:%5B%7B%22refId%22:%22B%22,%22expr%22:%22%7Bcluster%3D%5C%224am%5C%22,%20namespace%3D%5C%22qa%5C%22,%20pod%3D%5C%22level-zero-mem-1a-1664299480%5C%22%7D%20%7C%3D%20%5C%22ERROR%5C%22%20%21%3D%20%5C%22rejected%20by%20grpc%20RateLimiter%5C%22%20%7C%3D%20%5C%22timeout%5C%22%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22vhI6Vw67k%22%7D,%22editorMode%22:%22code%22,%22hide%22:false%7D%5D,%22range%22:%7B%22from%22:%221709755653916%22,%22to%22:%221710048176529%22%7D%7D%7D&schemaVersion=1&orgId=1)
![image](https://github.com/milvus-io/milvus/assets/27288593/6a9fba0c-61a1-4f07-939c-96878486aa0e)
- release name: level-zero-insert-op-39-9562

**Comment by xiaofan-luan** [comment]
seems to be much better

**Comment by ThreadDao** [comment]
@XuanYang-cn 
- image: 2.4-20240330-bc4a9a1ab-amd64
test: multi-partitions + concurrent: upsert, flush, search
https://argo-workflows.zilliz.cc/archived-workflows/qa/79bd882e-1767-4758-9269-6fbc7bd66730?nodeId=level-zero-stable-pnwdl

**Comment by XuanYang-cn** [comment]
It happens more easier when flush is not that frequently. Cuz next flush would seal the growing segments and advance cp.

**Comment by ThreadDao** [comment]
Fixed `cardinal-milvus-io-2.4-088455c-20240408`
Very cool. In the four scenarios of concurrent reading and writing, none of the flush requests timedout.
https://argo-workflows.zilliz.cc/archived-workflows/qa/e167ccf6-fe30-4933-bebe-391f0d493550?nodeId=level-zero-stable-5nklh
